src/checks/duplicate-payload-template-fields.ts

```diff
@@ -9,28 +9,12 @@ export const mustNotHaveDuplicateFieldNamesAfterEvaluation: AslChecker = (defini
     // for each one, examine each node's children for field names
     // that will be in conflict after evaluation.
     const errorMessages: StateMachineError[] = [];
-    let valuesToInspect: Record<string, unknown>[] = [];
-    try {
-        valuesToInspect = JSONPath<Record<string, unknown>[]>({
-            json: definition,
-            // this uses some extensions to the original JSONPath spec supported by the lib
-            //
-            // extension 1: @property.match enables matching on the matched property name.
-            //              This finds and selects all nodes with a name ending in .$
-            // extension 2: ^ operator twice at the end selects the parent node instead of the value for the node
-            //
-            // also, don't miss the array offset [0] before selecting the grandparent.
-            // There may be multiple fields within a node that could be in conflict.
-            // We're checking all the keys in the parent node, so we only need a single
-            // result for that parent.
-            path: `$..[Parameters,ResultSelector]..*[?(@property.match(/^.+\\.\\$$/))][0]^^`
-        })
```

Comment:
The expression (when it worked), identified all of the objects containing a key with a .$ suffix. This reduced the number of objects to inspect to just the ones that could be in conflict.

This expression didn't handle array types in the payload, perhaps due to a bug in jsonpath-plus. The descendants operator (..) and predicate with `property.match` throws if it encounters an array. The desired behavior IMHO would be to traverse the array's elements but that's not what the lib does.

Even if it did work, it's using extensions that aren't as widely known or documented. A simple imperative traversal like below is better than a declarative approach requiring all of these extra comments.